ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,906 @@
+            .map_err(ProverError::EvalError)
+            .map_err(|e| TxSigningError::ProverError(e, i))?;
+
+        let sigma_tree = reduction_result.sigma_prop;
+        hints_bag.add_hints_for_input(i, generate_commitments_for(sigma_tree, public_keys));
+    }
+    Ok(hints_bag)
+}
+
+/// Extracting hints from a transaction and outputs it's corresponding TransactionHintsBag
+pub fn extract_hints(
+    tx: &Transaction,
+    state_context: &ErgoStateContext,
+    boxes_to_spend: TxIoVec<ErgoBox>,
+    data_boxes: Option<TxIoVec<ErgoBox>>,
+    real_secrets_to_extract: Vec<SigmaBoolean>,
+    simulated_secrets_to_extract: Vec<SigmaBoolean>,
+) -> Result<TransactionHintsBag, TxSigningError> {
+    let mut hints_bag = TransactionHintsBag::empty();
+    for (i, input) in tx.inputs.iter().enumerate() {
+        let input_box = boxes_to_spend
+            .iter()
+            .find(|b| b.box_id() == input.box_id)
+            .ok_or(TxSigningError::InputBoxNotFound(i))?;
+        let height = state_context.pre_header.height;
+        let self_box = boxes_to_spend
+            .iter()
+            .find(|b| b.box_id() == tx.inputs.as_vec()[i].box_id)
+            .cloned()
+            .ok_or_else(|| {
+                TxSigningError::ContextError("self_index is out of bounds".to_string())
+            })?;
+        let outputs = tx
+            .output_candidates
+            .iter()
+            .enumerate()
+            .map(|(idx, b)| ErgoBox::from_box_candidate(b, tx.id(), idx as u16))
+            .collect::<Result<Vec<ErgoBox>, SigmaSerializationError>>()?;
+
+        let outputs_ir = outputs.into_iter().map(Rc::new).collect();
+        let data_inputs_ir = if let Some(data_inputs) = tx.data_inputs.as_ref() {
+            Some(data_inputs.clone().enumerated().try_mapped(|(idx, di)| {
+                data_boxes
+                    .as_ref()
+                    .ok_or(TxSigningError::DataInputBoxNotFound(idx))?
+                    .iter()
+                    .find(|b| di.box_id == b.box_id())
+                    .map(|b| Rc::new(b.clone()))
+                    .ok_or(TxSigningError::DataInputBoxNotFound(idx))
+            })?)
+        } else {
+            None
+        };
+        let inputs_ir = tx
+            .inputs_ids()
+            .clone()
+            .enumerated()
+            .try_mapped(|(idx, u)| {
+                boxes_to_spend
+                    .iter()
+                    .find(|b| u == b.box_id())
+                    .map(|b| Rc::new(b.clone()))
+                    .ok_or(TxSigningError::InputBoxNotFound(idx))
+            })?;
+        let extension = tx
+            .inputs
+            .get(i)
+            .ok_or_else(|| {
+                TxSigningError::ContextError(
+                    "self_index not found in spending transaction inputs".to_string(),
+                )
+            })?
+            .spending_proof
+            .extension
+            .clone();
+        let ctx = Rc::new(Context {
```

Comment:
Why not use `signing::make_context` here?

Reply:
We talk about the problem that occurs I manually create `Context` for now